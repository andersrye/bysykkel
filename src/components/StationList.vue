<template>
  <div>
    <StationListItem
      v-for="station in stationList"
      :station="station"
      :station-status="stationStatusById[station.station_id]"
      :key="station.station_id"
      @click="$emit('item-click', station.station_id)"
    />
  </div>
</template>

<script setup>
import {computed} from "vue";
import StationListItem from "@/components/StationListItem";

const props = defineProps({
  stationInfo: { type: Object, default: () => ({}) },
  stationStatus: { type: Object, default: () => ({}) },
  filterText: { type: String, default: "" }
})

defineEmits(['item-click'])

const stationStatusById = computed(() => {
  return props.stationStatus?.data?.stations?.reduce((acc, status) => {
    acc[status.station_id] = status
    return acc
  }, {}) ?? {}
})

const stationList = computed(() => {
  const stations = props.stationInfo?.data?.stations ?? []
  return stations
      .filter(station => {
        return !props.filterText
            || station.name?.toLowerCase()?.includes(props.filterText.toLowerCase())
            || station.address?.toLowerCase()?.includes(props.filterText.toLowerCase())
      })
      .sort((a, b) => a.name.localeCompare(b.name))
})
</script>
