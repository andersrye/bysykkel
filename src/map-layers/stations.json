{
  "id": "stations",
  "type": "circle",
  "source": "stations",
  "paint": {
    "circle-radius": 3,
    "circle-color": "#555"
  }
}
