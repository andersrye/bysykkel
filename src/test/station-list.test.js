import { shallowMount } from '@vue/test-utils'
import StationList from '../components/StationList.vue'
import { expect, test } from 'vitest'
import stations from './test-data/stations-combined.json'


describe('StationList', () => {

    test('mount component', async () => {
        const wrapper = shallowMount(StationList, {
            props: {stations: stations},
        })
        expect(wrapper.html()).toMatchSnapshot()
        expect(wrapper.findAllComponents({name: 'StationInfo'}).length).toBe(stations.length)
    })

    test('search should be case insensitive', async () => {
        const wrapper = shallowMount(StationList, {
            props: {stations: stations},
        })
        await wrapper.setProps({filterText: "Bekken"})
        expect(wrapper.findAllComponents({name: 'StationInfo'}).length).toBe(1)
        expect(wrapper.getComponent({name: 'StationInfo'}).props().station).toMatchObject({name: 'TÃ¸yenbekken'})
    })

    test('search should also match address', async () => {
        const wrapper = shallowMount(StationList, {
            props: {stations: stations},
        })
        await wrapper.setProps({filterText: "gata"})
        expect(wrapper.findAllComponents({name: 'StationInfo'}).length).toBe(1)
        expect(wrapper.getComponent({name: 'StationInfo'}).props().station).toMatchObject({address: 'Ringgata 2A'})
    })
})
