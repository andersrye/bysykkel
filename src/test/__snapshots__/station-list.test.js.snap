// Vitest Snapshot v1

exports[`StationList > mount component 1`] = `
"<div data-v-e8df3e1c=\\"\\">
  <div class=\\"box item-container\\" data-v-e8df3e1c=\\"\\">
    <station-info-stub station=\\"[object Object]\\" data-v-e8df3e1c=\\"\\"></station-info-stub>
  </div>
  <div class=\\"box item-container\\" data-v-e8df3e1c=\\"\\">
    <station-info-stub station=\\"[object Object]\\" data-v-e8df3e1c=\\"\\"></station-info-stub>
  </div>
  <div class=\\"box item-container\\" data-v-e8df3e1c=\\"\\">
    <station-info-stub station=\\"[object Object]\\" data-v-e8df3e1c=\\"\\"></station-info-stub>
  </div>
</div>"
`;
